content/blog/architecture-as-code-microservices/index.md

```diff
@@ -0,0 +1,105 @@
+
+Microservices deployed as stacks communicate with each other through the StackReference resource, which makes stack exports such as networking, environmental variables, and even Kubernetes configuration available to stacks in the project. The `StackReference` constructor takes as input a string of the form `<organization>/<project>/<stack>`, and lets services access the outputs of that stack. The following section walks through a microservices example that deploys a database on AWS RDS, a REST application in AWS Fargate, and an AWS Application Load Balancer.
+
+## A Microservices Application Infrastructure
+
+We'll use the [AWS Stack Reference example](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture) to illustrate how Pulumi uses Architecture as Code to create reusable components and deployment frameworks. We’ve previously [blogged]({{< relref "blog/architect-aws-application-infra-with-pulumi-stack-references" >}}) about this example and covered the code in some detail.
+
+Let's examine how it creates reusable components for building out resources. First up is the [VPC class](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) in the [networking stack](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture/networking).
+
+In the networking service, [`vpc.ts](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/networking/src/vpc.ts) creates a TypeScript class that extends Pulumi ComponentResource. A ComponentResource abstracts one or more children that do not require custom create, read, update, and delete operations for provisioning. We can add the related resources to the correct parent to build out the class.
+
+```ts
+export class Vpc extends ComponentResource {
+    vpc: aws.ec2.Vpc;
+    privateZone: aws.route53.Zone;
+    dhcpOptionSet: aws.ec2.VpcDhcpOptions;
+    internetGateway: aws.ec2.InternetGateway;
+    publicSubnets: aws.ec2.Subnet[] = [];
+    privateSubnets: aws.ec2.Subnet[] = [];
+    publicRouteTable: aws.ec2.RouteTable;
+    privateRouteTables: aws.ec2.RouteTable[] = [];
+    natGateways: aws.ec2.NatGateway[] = [];
+    natElasticIpAddresses: aws.ec2.Eip[] = [];
+
+   // ...
+
+`"
+
+The resource needs an interface to set the name and parameters for our new VPC class.
+
+```ts
+export interface VpcArgs {
+    description: string;
+    baseTags: aws.Tags;
+
+    baseCidr: string;
+    availabilityZoneNames: string[];
+    enableFlowLogs?: boolean;
+
+    endpoints: {
+        s3: boolean;
+        dynamodb: boolean;
+    };
+}
+```
+
+Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of Typescript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
```

Comment:
Nit:

```suggestion
Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of TypeScript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
```